File: web/src/lib/api/gen.ts
import { http, httpClient } from '@/lib/http'
import type {
  DbTableInfo,
  GenColumn,
  GenFile,
  GenTable,
  GenTableDetail,
  PageResult,
} from '@/lib/api/types'

export interface GenTableQuery {
  page?: number
  page_size?: number
  table_name?: string
}

export interface UpdateColumnPayload {
  id: string
  column_comment?: string
  is_required?: boolean
  is_insert?: boolean
  is_edit?: boolean
  is_list?: boolean
  is_query?: boolean
  sort?: number
}

export interface UpdateGenTablePayload {
  class_name?: string
  module_name?: string
  function_name?: string
  remark?: string | null
  columns?: UpdateColumnPayload[]
}

export const genApi = {
  dbTables: () => http.get<DbTableInfo[]>('/gen/db-tables'),
  list: (query: GenTableQuery) =>
    http.get<PageResult<GenTable>>('/gen/tables', query),
  import: (table_names: string[]) =>
    http.post<number>('/gen/import', { table_names }),
  detail: (id: string) => http.get<GenTableDetail>(`/gen/tables/${id}`),
  update: (id: string, payload: UpdateGenTablePayload) =>
    http.put<GenTableDetail>(`/gen/tables/${id}`, payload),
  remove: (id: string) => http.delete<null>(`/gen/tables/${id}`),
  preview: (id: string) => http.get<GenFile[]>(`/gen/tables/${id}/preview`),
  download: async (id: string, filename: string) => {
    const resp = await httpClient.get(`/gen/tables/${id}/download`, {
      responseType: 'blob',
    })
    const url = URL.createObjectURL(resp.data as Blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `${filename}.zip`
    a.click()
    URL.revokeObjectURL(url)
  },
}

export type { GenColumn }

File: crates/service/src/dto/gen.rs
use serde::{Deserialize, Serialize};
use validator::Validate;

use super::{de_u64, default_page, default_page_size, PageQuery};

#[derive(Debug, Deserialize, utoipa::IntoParams)]
#[into_params(parameter_in = Query)]
pub struct GenTableQuery {
    #[serde(default = "default_page", deserialize_with = "de_u64")]
    pub page: u64,
    #[serde(default = "default_page_size", deserialize_with = "de_u64")]
    pub page_size: u64,
    pub table_name: Option<String>,
}

impl GenTableQuery {
    pub fn pagination(&self) -> PageQuery {
        PageQuery::new(self.page, self.page_size)
    }
}

/// A physical table available in the database for import.
#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct DbTableInfo {
    pub table_name: String,
    pub comment: String,
    /// Whether this table has already been imported into the generator.
    pub imported: bool,
}

#[derive(Debug, Deserialize, Validate, utoipa::ToSchema)]
pub struct ImportTablesReq {
    #[validate(length(min = 1))]
    pub table_names: Vec<String>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct GenColumnReq {
    #[serde(with = "entity::id::string")]
    #[schema(value_type = String)]
    pub id: i64,
    pub column_comment: Option<String>,
    pub is_required: Option<bool>,
    pub is_insert: Option<bool>,
    pub is_edit: Option<bool>,
    pub is_list: Option<bool>,
    pub is_query: Option<bool>,
    pub sort: Option<i32>,
}

#[derive(Debug, Deserialize, Validate, utoipa::ToSchema)]
pub struct UpdateGenTableReq {
    #[validate(length(min = 1, max = 128))]
    pub class_name: Option<String>,
    #[validate(length(min = 1, max = 64))]
    pub module_name: Option<String>,
    #[validate(length(min = 1, max = 64))]
    pub function_name: Option<String>,
    #[validate(length(max = 255))]
    pub remark: Option<String>,
    #[serde(default)]
    pub columns: Vec<GenColumnReq>,
}

#[derive(Debug, Serialize, utoipa::ToSchema)]
pub struct GenTableDetail {
    pub table: entity::gen_table::Model,
    pub columns: Vec<entity::gen_column::Model>,
}

/// A single generated source file.
#[derive(Debug, Clone, Serialize, utoipa::ToSchema)]
pub struct GenFile {
    pub path: String,
    pub language: String,
    pub content: String,
}
